File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    private int n, cnt;
    
    public int solution(int[] numbers, int target) {
        n = numbers.length;
        cnt = 0;        
        dfs(numbers, 0, 0, target);
        
        return cnt;
    }
    
    private void dfs(int[] numbers, int depth, int sum, int target) {
        if (depth == n) {
            if (sum == target) cnt++;
            return;
        }
        
        dfs(numbers, depth + 1, sum + numbers[depth], target);
        dfs(numbers, depth + 1, sum - numbers[depth], target);
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.kt
class Solution {
    private var answer = 0

    fun solution(numbers: IntArray, target: Int): Int {
        dfs(-1, 0, numbers, target)
        return answer
    }

    private fun dfs(cnt: Int, sum: Int, numbers: IntArray, target: Int) {
        if(cnt == numbers.size - 1) {
            if(sum == target) answer++
            return
        }

        dfs(cnt+1, sum + numbers[cnt+1], numbers, target)
        dfs(cnt+1, sum - numbers[cnt+1], numbers, target)
    }
}